test/cmocka/src/math/dct/mfcc_get_dct_matrix.m

```diff
@@ -0,0 +1,53 @@
+% dct_matrix = mfcc_get_dct_matrix(num_ceps, num_mels, type)
+%
+% Input
+%   num_ceps - number of cepstral coefficients
+%   num_mels - number of Mel band energies
+%   type - DCT type, currently only 2 for DCT-II is supported
+%
+% Output
+%   dct matrix coefficients
+
+% SPDX-License-Identifier: BSD-3-Clause
+%
+% Copyright (c) 2022, Intel Corporation. All rights reserved.
+
+% Ref: pytorch kaldi.py _get_dct_matrix
+function dct_matrix = mfcc_get_dct_matrix(num_ceps, num_mels, type)
+
+dct_matrix = create_dct(num_ceps, num_mels, type, 'ortho');
+
+% make first cepstral to be weighted sum of factor sqrt(1 / num_mels)
+%dct_matrix(:,1) = sqrt(1 / num_mels);
+%dct_matrix = dct_matrix(:, 1:num_ceps);
+
+end
+
+
+% Ref pytorch functional.py create_dct
+function dct = create_dct(n_mfcc, n_mels, type, norm)
+
+n = (0:(n_mels - 1))'; % Size n_mels x 1
+k = 0:(n_mfcc - 1); % Size 1 x n_mfcc
+switch type
+	case 2,
+		arg = pi / n_mels * (n + 0.5) * k;
+		%max_arg = max(max(arg))
+		dct = cos(arg);
+	case 3,
+		error('not implemented'); % FIXME
+	otherwise
+		error('Unknown DCT type');
+end
+
+switch lower(norm)
+	case 'ortho',
+		dct(:,1) = dct(:,1) * 1 / sqrt(2);
```

Comment:
just the first column, why? Is it as complex as a dct matrix?

Reply:
I followed this implementation, where same things are done in Python arrays style: https://github.com/pytorch/audio/blob/95eada24fcff616aebdffe9aa0b174eee2264fc5/torchaudio/functional/functional.py#L626

The DCT-II description in Wikipedia mentions "Some authors further multiply the X_{0} term by 1/sqrt(2) and multiply the resulting matrix by an overall scale factor of sqrt(2/N) (see below for the corresponding change in DCT-III). This makes the DCT-II matrix orthogonal"

I looked up how orthogonality can be checked, in that case A x A^T = I, I'll check if that holds for this reference code.

Reply:
Thank you for the confirmation!
Why is there no orthogonality if the first column "dct(:,1)" is not selected?
I was still having trouble with the math. 

```
n_mels = 5;
n_mfcc = 3;
n = (0:(n_mels - 1))'; % Size n_mels x 1
k = 0:(n_mfcc - 1); % Size 1 x n_mfcc

arg = pi / n_mels * (n + 0.5) * k;
%max_arg = max(max(arg))
dct = cos(arg);

dct(:,3) = dct(:,3) * 1 / sqrt(2);
dct = dct * sqrt(2 / n_mels);

```

```
dct =

    0.6325    0.6015    0.3618
    0.6325    0.3717   -0.1382
    0.6325    0.0000   -0.4472
    0.6325   -0.3717   -0.1382
    0.6325   -0.6015    0.3618
    

```    
```
 dct' * dct

ans =

    2.0000   -0.0000         0
   -0.0000    1.0000         0
         0         0    0.5000
```

Reply:
That's would be deep mathematics, maybe this paper helps you to dig into why so https://math.mit.edu/~gs/papers/dct.pdf .

Reply:
Thank you very much in offering the explanation very significantly.
I tried to figure it out on my own, but the existing implementation is unaffected.

Reply:
My math is too rusty to understand now the paper but in this step it's enough that it computes the same as Pytorch. I will also make a configuration blob that matches Matlab's MFCC.